metadata/snapshot.go

```diff
@@ -302,14 +357,48 @@ func (s *snapshotter) createSnapshot(ctx context.Context, key, parent string, re
 	}
 
 	var (
-		target  = base.Labels[labelSnapshotRef]
-		bparent string
-		bkey    string
-		bopts   = []snapshots.Opt{
+		target               = base.Labels[labelSnapshotRef]
+		_, disableSameUnpack = base.Labels[labelDisableSameUnpack]
+		bparent              string
+		bkey                 string
+		bopts                = []snapshots.Opt{
 			snapshots.WithLabels(snapshots.FilterInheritedLabels(base.Labels)),
 		}
 	)
 
+	if disableSameUnpack {
+		bc, ok := s.inp.loadOrStore(key)
+		if ok {
+			// Wait for someone to broadcast that an active snapshot with the same key was either committed or removed.
+			err := bc.wait()
+			// If the extraction snapshot we were waiting on wasn't removed, then it either
+			// 1. Succeeded and we return ErrAlreadyExists
+			// 2. Failed and we return the error as is.
+			if !bc.remove {
```

Comment:
if I read this correctly and given what you described, if the first snapshot fails, then optimization is basically skipped for the snapshots that were waiting?

Reply:
So there's a couple failure modes I'd had in mind. 
1. If the diff being applied failed, then one of the waiters should be notified and they can proceed and retry (which isn't happening right now, they all get notified as I'm using broadcast everywhere). Then any of the other waiters simply stay blocked in wait() until someone succeeds.
2. The commit of the snapshot failed, in which case this is likely a symptom of a larger problem so we just broadcast the error to everyone and they should eventually bubble up and all fail.

Reply:
Need to fix the first point.

Reply:
I see. I guess you could just `Signal()`? not sure if this needs to be fixed now or can be optimized later. Depends on how urgent this needs to get in.

Reply:
Right, `Signal` should only notify one waiter which is ideal if there's more than one waiting as otherwise if something failed we fall back to the original behavior of everyone going to unpack the same layer.

Reply:
It's worth it to fix it I think

Reply:
if this resolves major use case, I'm ok with handling the 1st later.